test/tree.c

```diff
@@ -405,14 +359,636 @@ bool test_src_addr()
+			return false;
+		}
+	} else {
+		if (candidate_ctrl == reference_ctrl) {
+			printf("%s-%d-%d: Candidate (%s, %s, %s, %s, %s) should not match (%s, %s, %s, %s, %s, %s)\n",
+			       tag, reference_id, candidate_id,
+			       candidate->transport, candidate->traddr, candidate->trsvcid,
+			       candidate->host_traddr, candidate->host_iface,
+			       reference->transport, reference->traddr, reference->trsvcid,
+			       reference->host_traddr, reference->host_iface, reference->address);
+			return false;
+		}
+	}
+
+	/* Set the faked data back to NULL before freeing the tree */
+	reference_ctrl->name = NULL;
+	reference_ctrl->address = NULL;
+
+	nvme_free_tree(r);
+
+	return true;
+}
+
+/**
+ * test_ctrl_match_fc - Test that we can look up FC controllers
+ *
+ * @return true when all tests have passed. false otherwise.
+ */
+static bool test_ctrl_match_fc(void)
+{
+	bool pass = true;
+	struct ctrl_args reference = {0};
+	struct ctrl_args candidate = {0};
+
+	printf("test_ctrl_match_fc:\n");
+
+	/*******************************************************************/
+	/* Reference ID 1 */
+	set_ctrl_args(&reference, "fc", "21:00:00:e0:8b:05:05:01", "4420", "21:00:00:e0:8b:05:05:20", NULL, NULL);
+
+	set_ctrl_args(&candidate, "fc", "21:00:00:e0:8b:05:05:01", "4420", "21:00:00:e0:8b:05:05:20", NULL, NULL);
+	pass &= ctrl_match("FC", 1, 0, &reference, &candidate, true);
+
+	set_ctrl_args(&candidate, "fc", "21:00:00:e0:8b:05:05:01", "4420", NULL, NULL, NULL);
+	pass &= ctrl_match("FC", 1, 1, &reference, &candidate, true);
+
+	set_ctrl_args(&candidate, "fc", "21:00:00:e0:8b:05:05:01", "4420", NULL, "eth0", NULL);
+	pass &= ctrl_match("FC", 1, 2, &reference, &candidate, true);
+
+	set_ctrl_args(&candidate, "fc", "192.168.2.2", "4420", NULL, "eth0", NULL);
+	pass &= ctrl_match("FC", 1, 3, &reference, &candidate, false);
+
+	set_ctrl_args(&candidate, "fc", "21:00:00:e0:8b:05:05:01", "8009", NULL, NULL, NULL);
+	pass &= ctrl_match("FC", 1, 4, &reference, &candidate, false);
+
+	set_ctrl_args(&candidate, "fc", "21:00:00:e0:8b:05:05:01", "4420", "21:00:00:e0:8b:05:05:21", NULL, NULL);
+	pass &= ctrl_match("FC", 1, 5, &reference, &candidate, false);
+
+	set_ctrl_args(&candidate, "fc", "21:00:00:e0:8b:05:05:01", "8009", NULL, "", NULL);
+	pass &= ctrl_match("FC", 1, 6, &reference, &candidate, false);
+
+
+	/*******************************************************************/
+	/* Reference ID 2 */
+	set_ctrl_args(&reference, "fc", "21:00:00:e0:8b:05:05:01", "4420", NULL, NULL, NULL);
+
+	set_ctrl_args(&candidate, "fc", "21:00:00:e0:8b:05:05:01", "4420", "21:00:00:e0:8b:05:05:20", NULL, NULL);
+	pass &= ctrl_match("FC", 2, 0, &reference, &candidate, true);
+
+	set_ctrl_args(&candidate, "fc", "21:00:00:e0:8b:05:05:01", "4420", NULL, NULL, NULL);
+	pass &= ctrl_match("FC", 2, 1, &reference, &candidate, true);
+
+	set_ctrl_args(&candidate, "fc", "21:00:00:e0:8b:05:05:01", "4420", NULL, "eth0", NULL);
```

Comment:
Do these FC tests make sense? I mean the values for `trsvcid` and `host_iface`. Obviously, the libnvme doesn't care about these values but the kernel will. This is just a minor issue. I'll go ahead and merge it. I've updated my old test case and played a bit with it, to get a better understanding what's happening.

Reply:
Thanks for reviewing my PR. I provided these FC tests to make sure things don't blow up when passing illegal values and/or parameters. 

BTW, this is one of the reasons I provided transport-specific code in my earlier PR (ref: https://github.com/linux-nvme/libnvme/pull/695/files#diff-f131da46870af4738f860f74d79b2b4b8f8afd9900d5ea8e7956a37550c56c32R1421). The intention was to make it clear which parameters apply to each transport type and only test those parameters.

Reply:
Ah okay. I was just a bit confused when reading through this section.